feat(wallet): txBuilder disable after submit

mirceahasegan opened the pull request:
# Context
ADP-2192
I want to know if the TxBuilder instance was already used to build and submit a transaction, so that I can avoid building and sending another transaction using the same instance, that still has the configurations (outputs/delegations/etc.) from the previously sent transaction.

# Proposed Solution
TxBuilder `build()` method throws `TxBuilderDisabled` exception, if a transaction built with the same TxBuilder instance was already successfully submitted.
TxBuilder can be reenabled by calling it's `forceEnable()` method, to allow reusing it's configuration if needed.

# Important Changes Introduced

rhyslbw commented on packages/wallet/src/TxBuilder/types.ts:
```diff
@@ -32,13 +32,23 @@ export class OutputValidationTokenBundleSizeError extends CustomError {
   }
 }
 
+export class TxBuilderDisabled extends CustomError {
```
```suggestion
export class TxAlreadySubmitted extends CustomError {
```

mirceahasegan replied:
Sounds much better, thanks! I will also add the `Error` suffix: `TxAlreadySubmittedError`

mirceahasegan replied:
Updated in: [d0bfbb2](https://github.com/input-output-hk/cardano-js-sdk/pull/434/commits/d0bfbb2dbae618fec7ec774c5aa6b12578e92f34)

rhyslbw commented on packages/wallet/src/TxBuilder/types.ts:
```diff
@@ -173,12 +188,18 @@ export interface TxBuilder {
   build(): Promise<MaybeValidTx>;
 
+  /** Force enable TxBuilder to be reused, even if a transaction with this configuration was successfully submitted. */
+  forceEnable(): TxBuilder;
```
How about offering a method that removes the no longer relevant parts of the transaction. Maybe `copy()`?

mkazlauskas replied:
I think the idea here is to be able to reuse the parts. Although not sure if there are any uses cases outside of tests. If we want a clean state then `reset()` could work.

rhyslbw replied:
> Although not sure if there are any uses cases outside of tests

If that's the case it'd be better to remove it from the builder and implement a test util to provide similar convenience

mirceahasegan replied:
Agree. It makes it unnecessarily complex. I will remove it. For the moment we don't need it in tests either 😄

mirceahasegan replied:
Done in [d0bfbb2](https://github.com/input-output-hk/cardano-js-sdk/pull/434/commits/d0bfbb2dbae618fec7ec774c5aa6b12578e92f34)

mkazlauskas commented on packages/wallet/src/TxBuilder/types.ts:
```diff
@@ -173,12 +188,18 @@ export interface TxBuilder {
   build(): Promise<MaybeValidTx>;
 
+  /** Force enable TxBuilder to be reused, even if a transaction with this configuration was successfully submitted. */
+  forceEnable(): TxBuilder;
```
Let's try to find a better term for this. 1 suggestion below, but not sure about this. Thoughts @rhyslbw ?
```suggestion
  isDestroyed: boolean;
  restore(): TxBuilder;
```

rhyslbw replied:
Suggestion here:
https://github.com/input-output-hk/cardano-js-sdk/pull/434#discussion_r969158433

mirceahasegan replied:
It would be nice to have a getter, but `isDestroyed` is not very accurate. It's still there, you can use most of its methods, only build() is restricted.

mkazlauskas commented on packages/wallet/src/TxBuilder/buildTx.ts:
```diff
@@ -91,13 +97,18 @@ export class ObservableWalletTxBuilder implements TxBuilder {
     this.#logger = logger;
   }
 
+  isDestroyed(): boolean {
```
Let's keep the terminology consistent, I like this one (from the error name) better 
```suggestion
  isSubmitted(): boolean {
```

mirceahasegan replied:
Updated here: [e2a9138](https://github.com/input-output-hk/cardano-js-sdk/pull/434/commits/e2a9138bb61064f99b1d37cfbabfdffa81a77fac)

mkazlauskas reviewed (approved):
Nice work! :rocket:

mirceahasegan commented:
@mkazlauskas , @rhyslbw , I only added a couple more negative tests for error propagation, in [88982d9](https://github.com/input-output-hk/cardano-js-sdk/pull/434/commits/88982d9820d61fe73b10dff32732f05042215cb7).
I will squash the commits after your approval.
Thank you

mirceahasegan commented:
Rebased on master and squashed the fixups in [4c26833](https://github.com/input-output-hk/cardano-js-sdk/pull/434/commits/4c26833b988b9bc87c577a99a671ae21a33e7088).
Also added 2 more unit tests for error handling:
- [it('can return a sign error'](https://github.com/input-output-hk/cardano-js-sdk/blob/feat/tx-builder-no-reuse-after-submit/packages/wallet/test/integration/buildTx.test.ts#L539)
- [it('can return a submit error')](https://github.com/input-output-hk/cardano-js-sdk/blob/feat/tx-builder-no-reuse-after-submit/packages/wallet/test/integration/buildTx.test.ts#L550)

rhyslbw commented on packages/wallet/test/integration/buildTx.test.ts:
```diff
@@ -459,6 +460,106 @@ describe('buildTx', () => {
+      }
+    });
+
+    it('can return a sign error', async () => {
```
```suggestion
    it('rejects if error is encountered during transaction finalization', async () => {
```

mirceahasegan replied:
fixed/squashed/force pushed in one go since the reviews have settled.
[169d79b](https://github.com/input-output-hk/cardano-js-sdk/pull/434/commits/169d79b1749f882aaa667c5cae5fe0fe65c2d639)

rhyslbw commented on packages/wallet/test/integration/buildTx.test.ts:
```diff
@@ -459,6 +460,106 @@ describe('buildTx', () => {
+      await expect(builtTx.sign()).rejects.toBe(signError);
+    });
+
+    it('can return a submit error', async () => {
```
```suggestion
    it('rejects if error is encountered during submission', async () => {
```

mirceahasegan replied:
fixed/squashed/force pushed in one go since the reviews have settled.
[169d79b](https://github.com/input-output-hk/cardano-js-sdk/pull/434/commits/169d79b1749f882aaa667c5cae5fe0fe65c2d639)

rhyslbw reviewed (approved):
Great work @mirceahasegan :tada:
